core/src/banking_simulation.rs

```diff
@@ -0,0 +1,616 @@
+            .expect("timed hashes");
+
+        let base_event_time = raw_base_event_time - warmup_duration;
+        let base_simulation_time = SystemTime::now();
+
+        let sender_thread = thread::Builder::new().name("solSimSender".into()).spawn({
+            let exit = exit.clone();
+
+            move || {
+                let (mut non_vote_count, mut non_vote_tx_count) = (0, 0);
+                let (mut tpu_vote_count, mut tpu_vote_tx_count) = (0, 0);
+                let (mut gossip_vote_count, mut gossip_vote_tx_count) = (0, 0);
+
+                info!("start sending!...");
+                let timed_batches_to_send = self.banking_trace_events.packet_batches_by_time.range(base_event_time..);
+                info!(
+                    "simulating banking trace events: {} out of {}, starting at slot {} (based on {} from traced event slot: {}) (warmup: -{:?})",
+                    timed_batches_to_send.clone().count(),
+                    self.banking_trace_events.packet_batches_by_time.len(),
+                    self.first_simulated_slot,
+                    {
+                        let raw_base_event_time: chrono::DateTime<chrono::Utc> = raw_base_event_time.into();
+                        raw_base_event_time.format("%Y-%m-%d %H:%M:%S.%f")
+                    },
+                    slot_before_next_leader_slot,
+                    warmup_duration,
+                );
+                let mut simulation_duration_since_base = Duration::default();
+                let (
+                    mut last_log_duration,
+                    mut last_tx_count,
+                    mut last_non_vote_count,
+                    mut last_tpu_vote_tx_count,
+                    mut last_gossip_vote_tx_count
+                ) = (Duration::default(), 0, 0, 0, 0);
+                for (&event_time, (label, batches_with_stats)) in timed_batches_to_send {
+                    let required_duration_since_base =
+                        event_time.duration_since(base_event_time).unwrap();
+
+                    // Busy loop for most accurate sending timings
+                    loop {
+                        if simulation_duration_since_base > required_duration_since_base {
+                            break;
+                        }
+                        let current_simulation_time = SystemTime::now();
+                        simulation_duration_since_base = current_simulation_time
+                            .duration_since(base_simulation_time)
+                            .unwrap();
+                    }
+
+                    let sender = match label {
+                        ChannelLabel::NonVote => &non_vote_sender,
+                        ChannelLabel::TpuVote => &tpu_vote_sender,
+                        ChannelLabel::GossipVote => &gossip_vote_sender,
+                        ChannelLabel::Dummy => unreachable!(),
+                    };
+                    sender.send(batches_with_stats.clone()).unwrap();
```

Comment:
why clone instead of lettting the sender thread take ownership of these batches?

just reviewing on github, so can't see the type - is this Arc-ed, or just a clone of actual packet batch?

Reply:
the useful `BTreeMap::range()` don't allow because it's not `range_into` or something like that.

However, I noticed that I can use `BTreeMap::split_off()`: 3a3131e2adfb8d858d64b2143236f7f0e11a9f2a

> just reviewing on github, so can't see the type - is this Arc-ed, or just a clone of actual packet batch?

fyi, this is Arc-ed.

Reply:
> However, I noticed that I can use `BTreeMap::split_off()`: [3a3131e](https://github.com/anza-xyz/agave/commit/3a3131e2adfb8d858d64b2143236f7f0e11a9f2a)

related to the above, i further improved simulation jitter: 5e77bd7fa0e0ec1a453485a7aca69013fc1540b2